src/coreclr/interpreter/compiler.cpp

```diff
@@ -5000,7 +5000,80 @@ static OpcodePeepElement peepBoxUnboxOpcodes[] = {
     { 10, CEE_ILLEGAL } // End marker
 };
 
-OpcodePeepElement peepTypeValueTypeOpcodes[] = {
+static OpcodePeepElement peepBoxBrFalseOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_BRFALSE },
+    { 10, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepBoxBrFalseSOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_BRFALSE_S },
+    { 7, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepBoxBrTrueOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_BRTRUE },
+    { 10, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepBoxBrTrueSOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_BRTRUE_S },
+    { 7, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepBoxIsInstOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_ISINST },
+    { 10, CEE_ILLEGAL } // End marker
+};
+
+static OpcodePeepElement peepBoxIsInstBrFalseOpcodes[] = {
+    { 0, CEE_BOX },
+    { 5, CEE_ISINST },
```

Comment:
My immediate reaction is "why isn't the preceding `BOX + ISINST` peep by itself good enough" when I see this, but I assume it will be more evident later on.

I would also be concerned that this creates an ordering dependency where we need to make sure the more complex peeps are scanned before the simpler one, despite the simpler one being declared first here.

Reply:
I have followed the peephole check design that @davidwrighton  has established. The way it works is that we first match a list of opcodes and then check the various conditions. The current implementation didn't include "dynamic" forward lookup. There can be a better way to handle the peeps and we can optimize it that way later, but the purpose of this change was to  fix failures caused by CEE_BOX with ByRefLike struct as an argument, not to optimize anything.

I have originally made this change without the peephole stuff, using a clone of the JIT code and having it in the CEE_BOX handling code, but then I've noticed that the peephole stuff already handles box/unbox and I didn't want to introduce a different mechanism for a similar thing.

Reply:
I've run into a similar situation with the RuntimeAsync where the pattern match scheme I came up with was non-ideal due to the variations on IL instructions that are extremely similar (like stloc, stloc_s, stloc.0, stloc.1, stloc.2). What I ended up with was to have the outer validation step validate that the length of the peep was safe, and then some of the instruction validation was done in the validity function. I'm not convinced its all that good though. Also, for that case I made a seperate peep handler object for use within call instructions. Which might make sense here too.